website/content/en/docs/best-practices/observability-best-practices.md

```diff
@@ -0,0 +1,255 @@
+---
+title: "Operator Observability Best Practices"
+linkTitle: "Observability Best practices"
+weight: 6
+description: This guide describes the best practices concepts for adding Observability to operators.
+---
+
+## Operator Observability Best Practices
+
+In this document we will outline what operators require in order to meet the "Deep Insights" capability level and provide best practices and examples for creating metrics, recording rules and alerts. It is based on the general guidelines in [Operator Capability Levels](https://sdk.operatorframework.io/docs/overview/operator-capabilities/).
+
+**Note:** For technical documentation of how to add metrics to your operator, please read the [Metrics](https://book.kubebuilder.io/reference/metrics.html) section of the Kubebuilder documentation.
+
+### Deep Insights capability level requirements
+
+1. **Health and Performance metrics** for all of the operator components - Implemented based on the guidelines below.
+2. **Metrics Documentation** - All metrics should have documentation.
+3. **Alerts** for when things are not working as expected for each of the operator's components - Implemented based on the guidelines below.
+4. **Alerts Runbooks** - Each alert MUST include a runbook_url annotation and a runbook.
+5. **Alerts and Metrics Tests** - E2E Testing for metrics and alerts and unit tests for alerts.
+6. **Events** - Custom Resources MUST emit custom events for the operations taking place.
+7. **Metering** - Operator leverages Operator Metering.
+
+   **Note:** Metering isn't mandatory at this point in order to meet the "Deep Insights" capability level requirments.
+
+### Metrics Guidelines
+
+1. Metrics `Help` message should be verbose, since it can be used to create auto generated documentation, Like its done here for example [KubeVirt metrics](https://github.com/kubevirt/kubevirt/blob/main/docs/metrics.md) and generated by [KubeVirt metrics doc generator](https://github.com/kubevirt/kubevirt/blob/main/tools/doc-generator/doc-generator.go).
+
+The `help` message should include the following details:
+- What does this metric measure
+- What is the unit of measurement. See the [Prometheus base units](https://prometheus.io/docs/practices/naming/#base-units) and [Understanding metrics types](https://prometheus.io/docs/tutorials/understanding_metric_types/#types-of-metrics)
+- What does the output mean.
+When creating a new metric or recording rule that reports a resource like a ‘pod’ or a ‘container’ name, please make sure that the `namespace` is included, in order for it to be uniquely identified. 
```

Comment:
A consumer can install the Operator in any namespace so how the author will define the namespace beforehand?

Reply:
@camilamacedo86 The namespace label is populated via service discovery in general. The exception would be if you report a metric for a resource that doesn't live in the same namespace as the operator. Also in recording rules you need to make sure to also include the `namespace` label in the calculation. See the "Note" that I added below this line.

Reply:
@sradco 
What do you mean by `The namespace label is populated via service discovery in general`?
Could you please provide an example?

Reply:
@camilamacedo86  The service discovery knows the namespace associated to the pod/container and usually they are reported as part of the metrics without the need to specify it in the metric. But there are a few exceptions that this is not automatic, like when creating a recording rule. A recording rule is reported like a metric but doesn't include a "Type" (like gauge/counter/etc.) and doesn't include all fields from the respected metrics that it is based on. In recording rules the developer should include the namespace specifically when there is a resource that can have the same name in different namespaces.

Let me know if this is clear and feel free to suggest a better wording for this.

Reply:
@sradco Do you mean [labelling `namespace` on metrics](https://prometheus.io/docs/practices/naming/#labels) instead of k8s resources?